File: Luca/app/src/main/java/de/culture4life/luca/ui/history/HistoryViewModel.java
package de.culture4life.luca.ui.history;

import static de.culture4life.luca.history.HistoryManager.createCsv;
import static de.culture4life.luca.ui.accesseddata.AccessedDataDetailFragment.KEY_ACCESSED_DATA_LIST_ITEM;
import static de.culture4life.luca.ui.history.HistoryFragment.NO_WARNING_LEVEL_FILTER;

import android.app.Application;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.core.util.Pair;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.TimeUnit;
import java.util.stream.Collectors;

import de.culture4life.luca.R;
import de.culture4life.luca.dataaccess.AccessedTraceData;
import de.culture4life.luca.dataaccess.DataAccessManager;
import de.culture4life.luca.history.CheckOutItem;
import de.culture4life.luca.history.DataSharedItem;
import de.culture4life.luca.history.HistoryItem;
import de.culture4life.luca.history.HistoryManager;
import de.culture4life.luca.history.MeetingEndedItem;
import de.culture4life.luca.ui.BaseViewModel;
import de.culture4life.luca.ui.ViewError;
import de.culture4life.luca.ui.ViewEvent;
import de.culture4life.luca.ui.accesseddata.AccessedDataFragment;
import de.culture4life.luca.util.TimeUtil;
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers;
import io.reactivex.rxjava3.core.Completable;
import io.reactivex.rxjava3.core.Maybe;
import io.reactivex.rxjava3.core.Observable;
import io.reactivex.rxjava3.schedulers.Schedulers;
import timber.log.Timber;


public class HistoryViewModel extends BaseViewModel {

    private static final int TAN_CHARS_PER_SECTION = 4;

    private final HistoryManager historyManager;
    private final DataAccessManager dataAccessManager;

    private final MutableLiveData<List<HistoryListItem>> historyItems = new MutableLiveData<>();
    private final MutableLiveData<ViewEvent<String>> tracingTanEvent = new MutableLiveData<>();
    private final MutableLiveData<ViewEvent<List<AccessedTraceData>>> newAccessedData = new MutableLiveData<>();

    private ViewError dataSharingError;

    public HistoryViewModel(@NonNull Application application) {
        super(application);
        historyManager = this.application.getHistoryManager();
        dataAccessManager = this.application.getDataAccessManager();
    }

    @Override
    public Completable initialize() {
        return super.initialize()
                .andThen(Completable.mergeArray(
                        historyManager.initialize(application),
                        dataAccessManager.initialize(application)
                ))
                .andThen(invokeHistoryUpdate())
                .andThen(invokeNewAccessedDataUpdate());
    }

    private Completable invokeHistoryUpdate() {
        return Completable.fromAction(() -> modelDisposable.add(updateHistoryItems()
                .subscribeOn(Schedulers.io())
                .subscribe(
                        () -> Timber.i("Updated history"),
                        throwable -> Timber.w("Unable to update history: %s", throwable.toString())
                )));
    }

    private Completable invokeNewAccessedDataUpdate() {
        return Completable.fromAction(() -> modelDisposable.add(showNewAccessedDataIfAvailable()
                .subscribeOn(Schedulers.io())
                .subscribe(
                        () -> Timber.i("Updated accessed data"),
                        throwable -> Timber.w("Unable to update accessed data: %s", throwable.toString())
                )));
    }

    @Override
    public Completable keepDataUpdated() {
        return Completable.mergeArray(
                super.keepDataUpdated(),
                observeHistoryChanges()
        );
    }

    private Completable observeHistoryChanges() {
        return historyManager.getNewItems()
                .throttleLatest(1, TimeUnit.SECONDS)
                .switchMapCompletable(historyItem -> updateHistoryItems());
    }

    private Completable showNewAccessedDataIfAvailable() {
        return dataAccessManager.getAccessedTraceDataNotYetInformedAbout()
                .toList()
                .filter(accessedTraceData -> !accessedTraceData.isEmpty())
                .flatMapCompletable(this::informAboutDataAccess);
    }

    private Completable informAboutDataAccess(List<AccessedTraceData> accessedTraceData) {
        return update(newAccessedData, new ViewEvent<>(accessedTraceData))
                .andThen(dataAccessManager.markAllAccessedTraceDataAsInformedAbout());
    }

    private Completable updateHistoryItems() {
        return loadHistoryItems()
                .toList()
                .flatMapCompletable(items -> update(historyItems, items));
    }

    private Observable<HistoryListItem> loadHistoryItems() {
        // get all available items
        Observable<HistoryItem> cachedItems = historyManager.getItems().cache();

        // find items that belong to each other
        Observable<Pair<HistoryItem, HistoryItem>> cachedStartAndEndItems = cachedItems
                .flatMapMaybe(endItem -> getHistoryStartItem(endItem, cachedItems)
                        .map(startItem -> new Pair<>(startItem, endItem)))
                .cache();

        // find items that don't belong to another item
        Observable<HistoryItem> cachedSingleItems = cachedStartAndEndItems
                .flatMap(historyItemPair -> Observable.just(historyItemPair.first, historyItemPair.second))
                .toList()
                .flatMapObservable(startAndEndItemsList -> cachedItems
                        .filter(historyItem -> !startAndEndItemsList.contains(historyItem)));

        // convert and merge items
        return Observable.mergeArray(
                cachedStartAndEndItems.flatMapMaybe(this::createHistoryViewItem),
                cachedSingleItems.flatMapMaybe(this::createHistoryViewItem)
        ).sorted((first, second) -> Long.compare(second.getTimestamp(), first.getTimestamp()));
    }

    /**
     * Attempts to find the start item for the specified end item.
     *
     * @param endItem      the item to find the matching start item for
     * @param historyItems all available items, sorted by timestamp (descending)
     */
    private Maybe<HistoryItem> getHistoryStartItem(@NonNull HistoryItem endItem, @NonNull Observable<HistoryItem> historyItems) {
        return Maybe.fromCallable(() -> {
            // get the matching start item type
            if (endItem.getType() == HistoryItem.TYPE_CHECK_OUT) {
                return HistoryItem.TYPE_CHECK_IN;
            } else if (endItem.getType() == HistoryItem.TYPE_MEETING_ENDED) {
                return HistoryItem.TYPE_MEETING_STARTED;
            } else {
                return null;
            }
        }).flatMap(startItemType -> historyItems.filter(historyItem -> historyItem.getType() == startItemType)
                .filter(historyItem -> historyItem.getTimestamp() < endItem.getTimestamp())
                .firstElement());
    }

    private Maybe<HistoryListItem> createHistoryViewItem(@NonNull Pair<HistoryItem, HistoryItem> historyItemPair) {
        return Maybe.zip(createHistoryViewItem(historyItemPair.first), createHistoryViewItem(historyItemPair.second), (start, end) -> {
            end.setTime(TimeUtil.getReadableDurationWithPlural(end.getTimestamp() - start.getTimestamp(), application).blockingGet());
            HistoryListItem merged = new HistoryListItem();
            merged.setTitle(end.getTitle());
            merged.setDescription(end.getDescription());
            merged.setAdditionalTitleDetails(end.getAdditionalTitleDetails());
            merged.setTime(application.getString(R.string.history_time_merged, start.getTime(), end.getTime()));
            merged.setTimestamp(end.getTimestamp());
            merged.setTitleIconResourceId(end.getTitleIconResourceId());
            merged.setRelatedId(end.getRelatedId());
            merged.setAccessedTraceData(end.getAccessedTraceData());
            merged.setPrivateMeeting(end.isPrivateMeeting());
            merged.setGuests(end.getGuests());
            return merged;
        });
    }

    private Maybe<HistoryListItem> createHistoryViewItem(@NonNull HistoryItem historyItem) {
        return Maybe.fromCallable(() -> {
            HistoryListItem item = new HistoryListItem();
            item.setTimestamp(historyItem.getTimestamp());
            item.setTime(application.getString(R.string.history_time, TimeUtil.getReadableTime(application, historyItem.getTimestamp())));
            item.setRelatedId(historyItem.getRelatedId());
            switch (historyItem.getType()) {
                case HistoryItem.TYPE_CHECK_IN: {
                    item.setTitle(historyItem.getDisplayName());
                    break;
                }
                case HistoryItem.TYPE_CHECK_OUT: {
                    CheckOutItem checkOutItem = (CheckOutItem) historyItem;

                    item.setTitle(checkOutItem.getDisplayName());

                    List<AccessedTraceData> accessedTraceData = dataAccessManager.getPreviouslyAccessedTraceData(checkOutItem.getRelatedId())
                            .toList().blockingGet();
                    boolean accessed = !accessedTraceData.isEmpty();
                    item.setAccessedTraceData(accessedTraceData);
                    item.setAdditionalTitleDetails(application.getString(R.string.history_check_out_details, checkOutItem.getRelatedId()));
                    item.setTitleIconResourceId(accessed ? R.drawable.ic_eye : R.drawable.ic_information_outline);

                    List<String> children = checkOutItem.getChildren();
                    if (children != null && !children.isEmpty()) {
                        String currentDescription = item.getDescription();
                        StringBuilder builder = new StringBuilder();
                        if (currentDescription != null) {
                            builder = builder.append(currentDescription)
                                    .append(System.lineSeparator());
                        }
                        String childrenCsv = createCsv(children);
                        builder = builder.append(application.getString(R.string.history_children_title, childrenCsv));
                        item.setDescription(builder.toString());
                    }
                    break;
                }
                case HistoryItem.TYPE_MEETING_STARTED: {
                    item.setTitle(application.getString(R.string.history_meeting_started_title));
                    break;
                }
                case HistoryItem.TYPE_MEETING_ENDED: {
                    MeetingEndedItem meetingEndedItem = (MeetingEndedItem) historyItem;
                    item.setTitle(application.getString(R.string.history_meeting_ended_title));
                    item.setPrivateMeeting(true);
                    item.setAdditionalTitleDetails(application.getString(R.string.history_check_out_details, item.getRelatedId()));
                    item.setTitleIconResourceId(R.drawable.ic_information_outline);
                    if (meetingEndedItem.getGuests().isEmpty()) {
                        item.setDescription(application.getString(R.string.history_meeting_empty_description));
                    } else {
                        String guestCsv = createCsv(meetingEndedItem.getGuests());
                        item.setDescription(application.getString(R.string.history_meeting_not_empty_description, guestCsv));
                        item.setGuests(meetingEndedItem.getGuests());
                    }
                    break;
                }
                case HistoryItem.TYPE_CONTACT_DATA_REQUEST: {
                    DataSharedItem dataSharedItem = (DataSharedItem) historyItem;
                    item.setTitle(application.getString(R.string.history_data_shared_title));
                    item.setAdditionalTitleDetails(application.getString(R.string.history_data_shared_description, dataSharedItem.getDays()));
                    item.setTitleIconResourceId(R.drawable.ic_information_outline);
                    break;
                }
                default: {
                    Timber.w("Unknown history item type: %d", historyItem.getType());
                    return null;
                }
            }
            return item;
        }).doOnError(throwable -> Timber.w("Unable to create history view item for %s: %s", historyItem, throwable.toString()))
                .onErrorComplete();
    }

    public void onShareHistoryRequested(int days) {
        Timber.d("onShareHistoryRequested() called with: days = [%s]", days);
        modelDisposable.add(application.getRegistrationManager()
                .transferUserData(days)
                .doOnSubscribe(disposable -> {
                    updateAsSideEffect(isLoading, true);
                    removeError(dataSharingError);
                })
                .map(String::toUpperCase)
                .map(tracingTan -> {
                    if (tracingTan.contains("-")) {
                        return tracingTan;
                    }
                    StringBuilder tracingTanBuilder = new StringBuilder(tracingTan);
                    for (int i = 1; i < tracingTan.length() / TAN_CHARS_PER_SECTION; i++) {
                        int hyphenPosition = tracingTanBuilder.lastIndexOf("-") + TAN_CHARS_PER_SECTION + 1;
                        tracingTanBuilder.insert(hyphenPosition, "-");
                    }
                    return tracingTanBuilder.toString();
                })
                .flatMapCompletable(tracingTan -> Completable.mergeArray(
                        update(tracingTanEvent, new ViewEvent<>(tracingTan)),
                        historyManager.addDataSharedItem(tracingTan, days)
                ))
                .doOnError(throwable -> {
                    dataSharingError = createErrorBuilder(throwable)
                            .withTitle(R.string.error_request_failed_title)
                            .withResolveAction(Completable.fromAction(() -> onShareHistoryRequested(days)))
                            .withResolveLabel(R.string.action_retry)
                            .build();
                    addError(dataSharingError);
                })
                .doFinally(() -> updateAsSideEffect(isLoading, false))
                .subscribeOn(Schedulers.io())
                .subscribe(
                        () -> Timber.d("Received data sharing TAN"),
                        throwable -> Timber.w("Unable to get data sharing TAN: %s", throwable.toString())
                ));
    }

    public static List<HistoryListItem> filterHistoryListItems(List<HistoryListItem> items, int warningLevelFilter) {
        if (warningLevelFilter != NO_WARNING_LEVEL_FILTER) {
            return items.stream()
                    .filter(item -> item.containsWarningLevel(warningLevelFilter))
                    .map(item -> filterAccessedTraceData(item, warningLevelFilter))
                    .collect(Collectors.toList());
        } else {
            return items;
        }
    }

    private static HistoryListItem filterAccessedTraceData(HistoryListItem item, int warningLevelFilter) {
        item.setAccessedTraceData(item.getAccessedTraceData().stream()
                .filter(traceData -> traceData.getWarningLevel() == warningLevelFilter)
                .collect(Collectors.toList()));
        return item;
    }

    private static List<AccessedTraceData> filterTraceData(List<AccessedTraceData> items, int warningLevelFilter) {
        if (warningLevelFilter != NO_WARNING_LEVEL_FILTER) {
            return items.stream()
                    .filter(item -> item.getWarningLevel() == warningLevelFilter)
                    .collect(Collectors.toList());
        } else {
            return items;
        }
    }

    public void onShowAccessedDataRequested() {
        onShowAccessedDataRequested(new ArrayList(), NO_WARNING_LEVEL_FILTER);
    }

    public void onShowAccessedDataRequested(@NonNull List<AccessedTraceData> accessedTraceData, int warningLevelFilter) {
        Bundle bundle = new Bundle();
        List<AccessedTraceData> filteredData = filterTraceData(accessedTraceData, warningLevelFilter);
        if (!filteredData.isEmpty()) {
            AccessedTraceData firstItem = filteredData.get(0);
            if (filteredData.size() == 1) {
                modelDisposable.add(dataAccessManager.createAccessDataListItem(filteredData.get(0))
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribeOn(Schedulers.io())
                        .subscribe(dataListItem -> {
                            if (isCurrentDestinationId(R.id.historyFragment)) {
                                bundle.putSerializable(KEY_ACCESSED_DATA_LIST_ITEM, dataListItem);
                                navigationController.navigate(R.id.action_historyFragment_to_accessedDataDetailFragment, bundle);
                            }
                        }));
            } else {
                bundle.putString(AccessedDataFragment.KEY_TRACE_ID, firstItem.getTraceId());
                bundle.putInt(HistoryFragment.KEY_WARNING_LEVEL_FILTER, warningLevelFilter);
                navigationController.navigate(R.id.action_historyFragment_to_accessedDataFragment, bundle);
            }
        }
    }

    public LiveData<List<HistoryListItem>> getHistoryItems() {
        return historyItems;
    }

    public LiveData<ViewEvent<String>> getTracingTanEvent() {
        return tracingTanEvent;
    }

    public LiveData<ViewEvent<List<AccessedTraceData>>> getNewAccessedData() {
        return newAccessedData;
    }

}

File: Luca/app/src/main/java/de/culture4life/luca/ui/history/HistoryViewModel.kt
package de.culture4life.luca.ui.history

import android.app.Application
import android.os.Bundle
import androidx.core.util.Pair
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import de.culture4life.luca.R
import de.culture4life.luca.checkin.CheckInManager
import de.culture4life.luca.dataaccess.AccessedTraceData
import de.culture4life.luca.dataaccess.DataAccessManager
import de.culture4life.luca.history.*
import de.culture4life.luca.network.NetworkManager
import de.culture4life.luca.ui.BaseViewModel
import de.culture4life.luca.ui.ViewError
import de.culture4life.luca.ui.ViewEvent
import de.culture4life.luca.ui.accesseddata.AccessedDataDetailFragment
import de.culture4life.luca.ui.accesseddata.AccessedDataFragment
import de.culture4life.luca.ui.accesseddata.AccessedDataListItem
import de.culture4life.luca.ui.history.HistoryListItem.*
import de.culture4life.luca.util.TimeUtil
import de.culture4life.luca.util.addTo
import de.culture4life.luca.util.toDateTime
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import io.reactivex.rxjava3.core.*
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.schedulers.Schedulers
import timber.log.Timber
import java.net.HttpURLConnection
import java.util.*
import java.util.concurrent.TimeUnit
import java.util.stream.Collectors

class HistoryViewModel(application: Application) : BaseViewModel(application) {

    private var historyManager: HistoryManager = this.application.historyManager
    private val dataAccessManager: DataAccessManager = this.application.dataAccessManager
    private var checkInManager: CheckInManager = this.application.checkInManager

    private val tracingTanEvent = MutableLiveData<ViewEvent<String>>()
    private val newAccessedData = MutableLiveData<ViewEvent<List<AccessedTraceData>>>()
    private var dataSharingError: ViewError? = null

    private val historyItems = MutableLiveData<List<HistoryListItem>>()

    override fun initialize(): Completable {
        return super.initialize()
            .andThen(
                Completable.mergeArray(
                    historyManager.initialize(application),
                    dataAccessManager.initialize(application)
                )
            )
            .andThen(invokeHistoryUpdate())
            .andThen(invokeNewAccessedDataUpdate())
    }

    private fun invokeHistoryUpdate(): Completable {
        return Completable.fromAction {
            modelDisposable.add(
                updateHistoryItems()
                    .subscribeOn(Schedulers.io())
                    .subscribe(
                        { Timber.i("Updated history") },
                        { Timber.w("Unable to update history: %s", it.toString()) }
                    )
            )
        }
    }

    override fun keepDataUpdated(): Completable {
        return Completable.mergeArray(
            super.keepDataUpdated(),
            observeHistoryChanges()
        )
    }

    private fun observeHistoryChanges(): Completable {
        return historyManager.newItems
            .throttleLatest(1, TimeUnit.SECONDS)
            .switchMapCompletable { updateHistoryItems() }
    }

    private fun updateHistoryItems(): Completable {
        return loadHistoryItems()
            .toList()
            .flatMapCompletable { items: List<HistoryListItem> -> update(historyItems, items) }
    }

    private fun loadHistoryItems(): Observable<HistoryListItem> {
        // get all available items
        val cachedItems = historyManager.items.cache()

        // find items that belong to each other
        val cachedStartAndEndItems = cachedItems
            .flatMapMaybe { endItem: HistoryItem ->
                getHistoryStartItem(endItem, cachedItems)
                    .map { startItem: HistoryItem -> Pair(startItem, endItem) }
            }
            .cache()

        // find items that don't belong to another item
        val cachedSingleItems = cachedStartAndEndItems
            .flatMap { historyItemPair: Pair<HistoryItem, HistoryItem> ->
                Observable.just(historyItemPair.first, historyItemPair.second)
            }
            .toList()
            .flatMapObservable { startAndEndItemsList: List<HistoryItem> ->
                cachedItems.filter { historyItem: HistoryItem -> !startAndEndItemsList.contains(historyItem) }
            }

        // convert and merge items
        return Observable.mergeArray(
            cachedStartAndEndItems
                .flatMapMaybe(this::createHistoryViewItem),
            cachedSingleItems
                .flatMapMaybe(this::createHistoryViewItem)
        ).sorted { item1, item2 -> item2.timestamp.compareTo(item1.timestamp) }
    }

    /**
     * Attempts to find the start item for the specified end item.
     *
     * @param endItem      the item to find the matching start item for
     * @param historyItems all available items, sorted by timestamp (descending)
     */
    private fun getHistoryStartItem(endItem: HistoryItem, historyItems: Observable<HistoryItem>): Maybe<HistoryItem> {
        return Single.just(endItem)
            .filter {
                it.type == HistoryItem.TYPE_CHECK_OUT || it.type == HistoryItem.TYPE_MEETING_ENDED
            }
            .flatMap {
                Maybe.fromCallable {
                    // get the matching start item type
                    if (endItem.type == HistoryItem.TYPE_CHECK_OUT) {
                        HistoryItem.TYPE_CHECK_IN
                    } else {
                        HistoryItem.TYPE_MEETING_STARTED
                    }
                }.flatMap { startItemType ->
                    historyItems.filter { historyItem: HistoryItem -> historyItem.type == startItemType }
                        .filter { historyItem: HistoryItem -> historyItem.timestamp < endItem.timestamp }
                        .firstElement()
                }
            }
    }

    private fun createHistoryViewItem(historyItemPair: Pair<HistoryItem, HistoryItem>): Maybe<HistoryListItem> {
        return Maybe.zip(createHistoryViewItem(historyItemPair.first), createHistoryViewItem(historyItemPair.second)) { start, end ->
            end.time = TimeUtil.getReadableDateTimeDifference(application, start.timestamp.toDateTime(), end.timestamp.toDateTime())
            end.time = application.getString(R.string.history_time_merged, start.time, end.time)
            if (start is CheckInListItem && end is CheckOutListItem) {
                end.isContactDataMandatory = start.isContactDataMandatory
            }
            end
        }
    }

    private fun createHistoryViewItem(historyItem: HistoryItem): Maybe<HistoryListItem> {
        return Single.just(historyItem)
            .filter { HistoryListItem.canHandle(it) }
            .flatMap {
                Maybe.fromCallable {
                    when (historyItem.type) {
                        HistoryItem.TYPE_CHECK_IN -> CheckInListItem(application, historyItem as CheckInItem)
                        HistoryItem.TYPE_CHECK_OUT -> CheckOutListItem(
                            application,
                            historyItem as CheckOutItem,
                            dataAccessManager
                        )
                        HistoryItem.TYPE_MEETING_STARTED -> MeetingStartedListItem(application, historyItem)
                        HistoryItem.TYPE_MEETING_ENDED -> MeetingEndedListItem(
                            application,
                            historyItem as MeetingEndedItem
                        )
                        else -> DataSharedListItem(
                            application,
                            historyItem as DataSharedItem
                        )
                    }
                }
            }
            .doOnError { Timber.w("Unable to create history view item for %s: %s", historyItem, it.toString()) }
            .onErrorComplete()
    }

    private fun invokeNewAccessedDataUpdate(): Completable {
        return Completable.fromAction {
            modelDisposable.add(
                showNewAccessedDataIfAvailable()
                    .subscribeOn(Schedulers.io())
                    .subscribe(
                        { Timber.i("Updated accessed data") },
                        { Timber.w("Unable to update accessed data: %s", it.toString()) }
                    )
            )
        }
    }

    private fun showNewAccessedDataIfAvailable(): Completable {
        return dataAccessManager.accessedTraceDataNotYetInformedAbout
            .toList()
            .filter { accessedTraceData: List<AccessedTraceData?> -> accessedTraceData.isNotEmpty() }
            .flatMapCompletable { accessedTraceData: List<AccessedTraceData> ->
                informAboutDataAccess(accessedTraceData)
            }
    }

    private fun informAboutDataAccess(accessedTraceData: List<AccessedTraceData>): Completable {
        return update(newAccessedData, ViewEvent(accessedTraceData))
            .andThen(dataAccessManager.markAllAccessedTraceDataAsInformedAbout())
    }

    fun onShareHistoryRequested(days: Int) {
        Timber.d("onShareHistoryRequested() called with: days = [%s]", days)
        modelDisposable.add(
            application.registrationManager
                .transferUserData(days)
                .doOnSubscribe {
                    updateAsSideEffect(isLoading, true)
                    removeError(dataSharingError)
                }
                .map { obj: String -> obj.uppercase(Locale.getDefault()) }
                .map { tracingTan: String ->
                    if (tracingTan.contains("-")) {
                        return@map tracingTan
                    }
                    val tracingTanBuilder = StringBuilder(tracingTan)
                    for (i in 1 until tracingTan.length / TAN_CHARS_PER_SECTION) {
                        val hyphenPosition = tracingTanBuilder.lastIndexOf("-") + TAN_CHARS_PER_SECTION + 1
                        tracingTanBuilder.insert(hyphenPosition, "-")
                    }
                    tracingTanBuilder.toString()
                }
                .flatMapCompletable { tracingTan: String ->
                    Completable.mergeArray(
                        update(tracingTanEvent, ViewEvent(tracingTan)),
                        historyManager.addDataSharedItem(tracingTan, days)
                    )
                }
                .doOnError { throwable: Throwable? ->
                    dataSharingError = createErrorBuilder(throwable!!)
                        .withTitle(R.string.error_request_failed_title)
                        .withResolveAction(Completable.fromAction { onShareHistoryRequested(days) })
                        .withResolveLabel(R.string.action_retry)
                        .build()
                    addError(dataSharingError)
                }
                .doFinally { updateAsSideEffect(isLoading, false) }
                .subscribeOn(Schedulers.io())
                .subscribe(
                    { Timber.d("Received data sharing TAN") },
                    { Timber.w("Unable to get data sharing TAN: %s", it.toString()) }
                )
        )
    }

    @JvmOverloads
    fun onShowAccessedDataRequested(
        accessedTraceData: List<AccessedTraceData> = ArrayList(),
        warningLevelFilter: Int = HistoryFragment.NO_WARNING_LEVEL_FILTER
    ) {
        val bundle = Bundle()
        val filteredData = filterTraceData(accessedTraceData, warningLevelFilter)
        if (filteredData.isEmpty()) {
            return
        }
        val firstItem = filteredData[0]
        if (filteredData.size == 1) {
            dataAccessManager.getNotificationTexts(firstItem)
                .map { AccessedDataListItem.from(getApplication(), firstItem, it) }
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe { dataListItem: AccessedDataListItem ->
                    if (isCurrentDestinationId(R.id.historyFragment)) {
                        bundle.putSerializable(AccessedDataDetailFragment.KEY_ACCESSED_DATA_LIST_ITEM, dataListItem)
                        navigationController!!.navigate(R.id.action_historyFragment_to_accessedDataDetailFragment, bundle)
                    }
                }
                .addTo(modelDisposable)
        } else {
            bundle.putString(AccessedDataFragment.KEY_TRACE_ID, firstItem.traceId)
            bundle.putInt(HistoryFragment.KEY_WARNING_LEVEL_FILTER, warningLevelFilter)
            navigationController!!.navigate(R.id.action_historyFragment_to_accessedDataFragment, bundle)
        }
    }

    fun onDeleteSelectedHistoryListItemsRequested() {
        Flowable.fromIterable(getHistoryItems().value!!)
            .filter { it.isSelectedForDeletion && !HistoryListItem.isContactDataMandatory(it) }
            .flatMapCompletable { item ->
                val relatedId = item.relatedId
                if (item is MeetingStartedListItem || item is MeetingEndedListItem) {
                    checkInManager.deleteCheckInLocally(relatedId)
                } else {
                    checkInManager.deleteCheckInFromBackend(relatedId)
                        .onErrorResumeNext { throwable: Throwable ->
                            if (NetworkManager.isHttpException(throwable, HttpURLConnection.HTTP_FORBIDDEN, HttpURLConnection.HTTP_NOT_FOUND)) {
                                // 403: incorrect signature, no way to recover
                                // 404: relatedId not found in backend, maybe already deleted
                                checkInManager.deleteCheckInLocally(relatedId)
                            } else {
                                Completable.error(throwable)
                            }
                        }
                }
            }
            .doOnError {
                val viewError = ViewError.Builder(application)
                    .withCause(it)
                    .removeWhenShown()
                addError(viewError.build())
            }
            .andThen(invokeHistoryUpdate())
            .subscribe(
                { Timber.v("Deleted selected history list items") },
                { Timber.w("Unable to delete historyListItems: %s", it.toString()) }
            )
    }

    fun getTracingTanEvent(): LiveData<ViewEvent<String>> = tracingTanEvent

    fun getNewAccessedData(): LiveData<ViewEvent<List<AccessedTraceData>>> = newAccessedData

    fun getHistoryItems(): LiveData<List<HistoryListItem>> = historyItems

    companion object {

        private const val TAN_CHARS_PER_SECTION = 4

        fun filterHistoryListItems(items: List<HistoryListItem>, warningLevelFilter: Int): List<HistoryListItem> {
            return if (warningLevelFilter != HistoryFragment.NO_WARNING_LEVEL_FILTER) {
                items.stream()
                    .filter { it is CheckOutListItem && it.containsWarningLevel(warningLevelFilter) }
                    .map { filterAccessedTraceData(it as CheckOutListItem, warningLevelFilter) }
                    .collect(Collectors.toList())
            } else {
                items
            }
        }

        private fun filterAccessedTraceData(item: CheckOutListItem, warningLevelFilter: Int): HistoryListItem {
            item.accessedTraceData = item.accessedTraceData.stream()
                .filter { traceData: AccessedTraceData -> traceData.warningLevel == warningLevelFilter }
                .collect(Collectors.toList())
            return item
        }

        private fun filterTraceData(items: List<AccessedTraceData>, warningLevelFilter: Int): List<AccessedTraceData> {
            return if (warningLevelFilter != HistoryFragment.NO_WARNING_LEVEL_FILTER) {
                items.stream()
                    .filter { item: AccessedTraceData -> item.warningLevel == warningLevelFilter }
                    .collect(Collectors.toList())
            } else {
                items
            }
        }
    }
}
